Focus submenu button when clicked

SantosGuillamot opened the pull request:
## What?
This pull request adds some JavaScript logic to ensure focus the submenu button when it is clicked. It solves this issue: https://github.com/WordPress/gutenberg/issues/54991

It also adds a fix for closing the submenu overlay on a click outside the submenu on iOS.

## Why?
It seems that Safari doesn't place the focus on the button element but on a parent `div`, which causes some issues with the directives assuming the focus is on the button. We can see the difference between Chrome and Safari:

**Chrome**

https://github.com/WordPress/gutenberg/assets/34552881/021bbbda-fb39-4b26-af8b-6dd4a504790b

**Safari**

https://github.com/WordPress/gutenberg/assets/34552881/2017fb96-f4d6-47f4-b433-568b473b3f7f

## How?
I added some JavaScript logic that, when the submenu button is clicked, it focuses it. It'd be great if someone from the accessibility team could double-check if this approach is correct or if it will break something I am unaware of. cc: @afercia, @alexstine , @jeryj , @joedolson  

## Testing Instructions
1. Add a navigation block with an item with submenu.
2. In the block settings, enable the "Show arrow" setting.
3. Open the site in Safari.
4. Click on the arrow to open the submenu.
5. Check that the `activeElement` is the arrow button.
6. Click outside of the submenu and check it is closed.
7. Open the submenu again by clicking the arrow.
8. Press the ESC key and check that the submenu is closed.

On iOS:
1. Click on the arrow to open the submenu.
2. Click on the arrow to close the submenu.
3. Click on the arrow to open the submenu.
4. Click on the body to close the submenu.

SantosGuillamot committed:
Focus element manually when open submenu on click

alexstine reviewed (approved):
@SantosGuillamot Thanks for the PR.

From what I can tell, this is at least a partial fix. I tested on Windows 10 with NVDA in Google Chrome and Firefox. I tested on Mac Big Sur with Voiceover and Safari. This seems like a solid fix across OS platforms.

I can however not test the click behavior with the mouse, this is only a test done with the keyboard. I would wait for one more approve on this one.

jeryj reviewed (changes requested):
This is testing well for keyboard interactions. The test steps also work correctly in safari with a mouse click. I noticed one regression from `trunk`. 

1. Open the site in Safari.
2. Click on the arrow to open the submenu.
4. Click _on the arrow to close the submenu_.
5. The submenu stays open, when it should close as it does on Trunk and in Chrome.

tellthemachines reviewed (approved):
Thanks for fixing this! It's working well for me on both Safari and the iPad simulator.

luisherranz commented:
Thanks a lot for testing this. 

We are trying to avoid changing the focus programmatically to avoid introducing bugs [like this](https://github.com/WordPress/gutenberg/pull/55198#pullrequestreview-1668520039) so we are now experimenting with a different fix that uses `tabindex="-1"` in the `<li>` element to make Safari move the focus to that `li` element instead of to the top container.

It seems to work fine, but before committing to it, we want to make sure that that behavior is consistent across all the different Safari versions and devices. We'll test it out with BrowserStack, and we'll let you know as soon as possible.

luisherranz committed:
Try using `tabindex="-1"`

luisherranz committed:
Use `tabindex="-1"` also in body when a submenu is opened

luisherranz commented:
@santosguillamot and I have been testing the `tabindex` fix using BrowserStack (real iPhone and iPad devices) and it looks like it fixes all the issues.

Could you please check again? Thank you 🙏

tellthemachines commented:
I'm not able to test this on Windows because gutenberg.run seems to be failing for this PR. Testing on macOS/Safari both mouse and keyboard interaction are working fine. I tested opening and closing submenus and nested submenus and don't think I missed anything this time 😄

jeryj commented on packages/block-library/src/navigation/view.js:
```diff
@@ -20,6 +20,9 @@ const openMenu = ( store, menuOpenedOn ) => {
 		document.documentElement.classList.add( 'has-modal-open' );
+	} else {
+		// Ensure that Safari on iOS/iPadOS trigger the mouseleave events.
+		document.body.setAttribute( 'tabindex', '-1' );
```
Do we know why this works? I'm not saying we can't do this, but I'm hesitant to do this since it could have far-reaching effects. I'd love to learn more about it as I didn't know this would trigger mouseleave events for Safari.

alexstine replied:
Yes, I don't like this. I would never set this on the `<body>`.

jeryj reviewed (changes requested):
Sorry to keep blocking this :/ 

The good news: It does work! I can't break it using TT3 theme. 
The odd news... the `tabindex` isn't always consistent in the value in relation to the state. I can get the menu to be open, and not have a `tabindex` value by:

- Tab to a dropdown chevron
- Press Enter: `tabindex` -1 is applied body tag
- Press Shift + Tab to go back to the chevron: `tabindex` is removed from the body tag

It can also arrive in the "wrong" state with hovering/clicks:

https://github.com/WordPress/gutenberg/assets/967608/772abf62-cd93-4e88-bc6a-289d8905fa1e


It also allows the focus to be placed on the `<li>` in some instances on Safari and Chrome. 

- Tab focus to a link with a dropdown
- Hover the link
- Hover off the link
- Focus ring is wrapped around both the anchor and chevron
- `document.activeElement` reports the `<li>`


https://github.com/WordPress/gutenberg/assets/967608/f9c0ff72-e03f-4b22-9b59-07949ce50aa1

Overall, I'm not sure what's going on with it, but it seems inconsistent. I'm really interested in why this fix works. Since I can get the body to not have a `tabindex` in Safari, I thought the bug of not being able to click the page to close the menu would still be there... But it isn't...  This code works to address that.

If we can isolate exactly what is allowing it to work in Safari, then we can use that to provide a fix that will work without potentially introducing odd states like the ones above.

So, sorry for the long text and for keeping this blocked. I'm super curious to find out more!

luisherranz commented:
The current implementation is coupled with how the focus is positioned in Chrome, so it's difficult to make it work in Safari because the focus position is totally different. Using `tabindex="-1"` was an attempt to get Safari to function in a manner similar to Chrome, but this may not be the best solution.

The problem is that the ideal solution would be to change the implementation so it doesn't depend on the focus position, but this would require a complete refactoring of the logic, and it might be a bit late in the WP 6.4 release cycle for such a big change.

We can keep trying to fix it for WP 6.4 with `tabindex` or `ref.focus()`, even if it's not ideal, and refactor the logic in Gutenberg, but I think it would be easier if we had direct help from an accessibility expert because I'm afraid we don't fully understand how `tabindex` is supposed to work, at least in relation to how it's expected to work in Chrome, why it's so different in Safari, and how it's expected to work in Safari.

@jeryj, would there be any chance we could pair program with you to see if we can find the best solution together?

alexstine commented:
@luisherranz I'll give +1 to just setting focus via the ref. That seemed to work.

luisherranz commented:
Thanks, Alex. I've talked with @jeryj and we'll meet tomorrow to keep iterating this 🙂

luisherranz committed:
Replace tabindex with event listener

joedolson commented:
Safari doesn't set focus on buttons, links, or non-text inputs; so `activeElement` won't reliably return a valid value if focus is expected on one of those items. Safari doesn't consider this a bug, it's a design choice, so it has to be worked around. I haven't looked at this PR yet, but that's an important weird behavior to be aware of with Safari. Relevant reading: [Fixing Focus for Safari](https://itnext.io/fixing-focus-for-safari-b5916fef1064)

luisherranz committed:
Explain the tabindex on <li>

luisherranz committed:
Don't store the element on hover to restore the focus later

luisherranz committed:
Improve explanations

luisherranz commented:
Ok, so this is the current status:

We replaced the `tabindex="-1"` of the `body` with a global event listener which has the same effect.

We tried to use `ref.focus()` manually instead of the `tabindex="-1"` in the `li` element, but it breaks the closing of the menu. To solve that problem, we would need to differentiate between a `focusout` because the user clicked on a button, and a `focusout` because the user clicked away on the body. But that seems impossible to do in Safari because there's nothing in the event information about what element had the focus previously (`event.relatedTarget` is `null` in both cases). Also, the `focusout` is the first event that is being triggered in Safari, so we can't set a flag in the `click` event that we can then read in the `focusout`. We tried delaying the `focusout` logic to force the `click` to go first with both `setTimeout` and `requestAnimationFrame`, but it doesn't seem reliable at all, especially in Safari.

So, we want to replace the current implementation with one based on overlays to avoid relying on `focusout`, but we don't feel like doing such a big change is a good idea with so little time to test it out before WP 6.4. So we think that using `tabindex="-1"` in the `li` element is the best temporary fix. We solved the problem with the focus ring moving to the `li` element (that was actually a bug, fixed in https://github.com/WordPress/gutenberg/pull/55198/commits/74525781a549b20fa04634ac0df10aa21926e15f), so now everything should be working fine and consistently across browsers.

We would appreciate another round of testing. Thanks 🙂🙏

jeryj committed:
Add tests to cover webkit frontend menu interactions

Safari doesn't place focus on a clicked button as expected. These tests verify that when a submenu chevron button is clicked, focus is correctly placed on that button. It also verifies that the click on the body correctly closes any open submenus, which was failing in Safari.

jeryj committed:
Focus clicked button on Safari

Combining the tabindex -1 on the parent li and focusing the button on Safari, and also checking that the relatedTarget is null inside the handleMenuFocusout seems to contain the focus within the menu to not fire the handleMenuFocusout as often, and still works to click on the body to close the menu.

jeryj commented:
> We tried to use ref.focus() manually instead of the tabindex="-1" in the li element, but it breaks the closing of the menu.

I played around with this too. I accidentally stumbled upon using _both tabindex="-1" in the li element AND using ref.focus()_ seems to work well. I think it's because the tabindex="-1" n the li prevents Safari from sending focus to the parent navigation wrapper. Combining this with a `event.relatedTarget === null` check on the `handleMenuFocusout` seems to prevent the submenus from being closed and also allows a click outside the menu to close the submenu. I didn't test on iPhone or iPad proper, so there's a chance all my work was for nothing 😅 

I added some checks to verify focus on the frontend interactivity tests as well, and added directives to run it in Safari. I spent the last hour debugging why it's flaky though. I believe it's an issue with the test set-up in Safari though. Checking it manually with the same steps consistently works. If we need to remove the Tab keypresses that keep breaking it or write them in a way that tests what we need to while being consistent, I'm fine with that. 

🤞🏻 it all works!

jeryj commented:
Tests pass, but they are flaky, as indicated by https://github.com/WordPress/gutenberg/issues/51552. To run the tests locally, you can do: `npm run test:e2e:playwright -- --project=webkit --repeat-each=10 -g "Submenu interactions"`. You may need to first do `npx playwright install webkit` to get the webkit browser to work with Playwright.

SantosGuillamot commented:
Thanks a lot for your help and the tests improvements 🙂 

> I didn't test on iPhone or iPad proper, so there's a chance all my work was for nothing 😅

I've been testing the latest changes in iPhone and iPad, and unfortunately, it doesn't seem to work there 😞 If I open a submenu, it is not closed by clicking outside. Adding the empty click event listener seems to fix the issue. Going back to Luis' fix everything seems to work fine as well.

jeryj committed:
Added the document.addEventListener body click back in

Authored by Luis Herranz <luisherranz@gmail.com>. I'm just re-applying the change.

jeryj committed:
Remove tab keypresses from webkit menu interaction tests

Tab keypreses on webkit playwright are really flakey (or it's something in our code that we haven't isolated) so I've split out the webkit tests to test everything I can without using a tab keypress.

jeryj commented:
I approved the changes since there wasn't a way for me to dismiss my own review. I think it needs a review from someone else now since I've made code updates.

jeryj committed:
Use body click instead for consistency across environments

luisherranz commented:
For reference, this is the implementation that was finally merged:

- We used a global event listener to force Safari in iOS/iPadOS to trigger the `focusout` event when a submenu is opened and the user taps away.
- We used a combination of `tabindex` and `ref.focus()` to manually position the focus on the button when the user clicks it. This only affects Safari. Chrome and Firefox keep working as they were.

So this should now be a fully working solution. 

We still want to investigate an alternative overlay-based solution at some point, though.

Thanks to everyone for their help with code and/or reviews 🙂

SiobhyB commented:
I've gone ahead to cherry-pick this change for inclusion in 6.4's RC2: https://github.com/WordPress/gutenberg/pull/55481/commits/042dfed013fdb7e809b9753a885b34564aa62450

Mamaduka commented:
I've noticed that `Overlay menu interactions in Safari` test introduced here become flaky recently - https://github.com/WordPress/gutenberg/issues/55816.

SantosGuillamot commented:
> I've noticed that Overlay menu interactions in Safari test introduced here become flaky recently - https://github.com/WordPress/gutenberg/issues/55816.

@jeryj Any idea why this could happen or how we can solve it?

krokodok commented:
There's a problem with closing the menu when clicking inside the debugger or address bar, stemming from this PR and described in this ticket: https://core.trac.wordpress.org/ticket/60192

Mamaduka commented:
@krokodok, a similar issue was reported recently. I don't think that's a bug; it's how most dialogs/popovers behave. See: https://github.com/WordPress/gutenberg/issues/56094#issuecomment-1809562491

krokodok commented:
Didn't know about that debugging option, good to know!

krokodok commented:
From my own personal view I think it is a benefit being able to open a state and then leaving the browser, maybe checking some related stuff on another monitor, and then returning to the browser and finding the exact same state that I left. When the state changes that feels sudden and confusing.

As I said, the implementations differ from website to website. Google for example, on their home search page, do it like it is done in WordPress right now, regarding the user profile menu.
